eva-remapping-ingest/src/main/java/uk/ac/ebi/eva/remapping/ingest/batch/io/RemappedSubmittedVariantsWriter.java

```diff
@@ -17,58 +17,327 @@
+
     private RemappingIngestCounts remappingIngestCounts;
 
-    public RemappedSubmittedVariantsWriter(MongoTemplate mongoTemplate, String collection,
+    public RemappedSubmittedVariantsWriter(MongoTemplate mongoTemplate, String assemblyAccession, String collection,
                                            RemappingIngestCounts remappingIngestCounts) {
         this.mongoTemplate = mongoTemplate;
+        this.assemblyAccession = assemblyAccession;
         this.collection = collection;
+        this.operationCollection = collection.equals(CollectionNames.SUBMITTED_VARIANT_ENTITY) ?
+                CollectionNames.SUBMITTED_VARIANT_OPERATION_ENTITY : CollectionNames.DBSNP_SUBMITTED_VARIANT_OPERATION_ENTITY;
         this.remappingIngestCounts = remappingIngestCounts;
     }
 
     @Override
     public void write(List<? extends SubmittedVariantEntity> submittedVariantsRemapped) {
-        try {
-            BulkOperations bulkOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
-                                                                  SubmittedVariantEntity.class,
-                                                                  collection);
-            bulkOperations.insert(submittedVariantsRemapped);
+        // We do this to avoid tedious generic type signatures in the method calls
+        List<SubmittedVariantEntity> svesToInsert = new ArrayList<>(submittedVariantsRemapped);
+
+        // Resolve duplicate hashes and accessions before inserting; note that hash resolution must happen first.
+        ImmutablePair<List<SubmittedVariantEntity>, List<SubmittedVariantOperationEntity>> duplicateHashDiscards =
+                resolveDuplicateHashes(svesToInsert);
+        ImmutablePair<List<SubmittedVariantEntity>, List<SubmittedVariantOperationEntity>> duplicateAccessionDiscards =
+                resolveDuplicateAccessions(svesToInsert);
+
+        List<SubmittedVariantEntity> svesToDiscard = Stream.concat(duplicateHashDiscards.getLeft().stream(),
+                                                                    duplicateAccessionDiscards.getLeft().stream())
+                                                            .collect(Collectors.toList());
+        List<SubmittedVariantOperationEntity> discardOperations = Stream.concat(duplicateHashDiscards.getRight().stream(),
+                                                                                duplicateAccessionDiscards.getRight().stream())
+                                                                        .collect(Collectors.toList());
+
+        BulkOperations bulkOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
+                                                              SubmittedVariantEntity.class,
+                                                              collection);
+        // Deal with discards before inserts, to avoid DuplicateKeyExceptions
+        if (svesToDiscard.size() > 0) {
```

Comment:
I am thinking if we should write the operations before removing SVEs. While that may seem to be counterintuitive: the line of reasoning here is:  op written -> sve removal fails -> sve removal redoable by consulting op. But consider the sequence: sve removed -> op write fails -> op write not redoable because the SVE will not be around in the svesToDiscard when the process is restarted. Is this a plausible scenario?

doing so helps us manually (or in this writer itself if we introduce some idempotency later) remove SVEs by consulting the operations. IOW, writing operations that are not truthful might be a lesser crime :)

Reply:
I think that makes sense, I can switch the order of these...